src/transports/smart.c

```diff
@@ -66,6 +66,76 @@ static int git_smart__set_callbacks(
 	return 0;
 }
 
+#define forbid_custom_header(disallowed_name) \
+	if (strncmp(disallowed_name, custom_header, name_len) == 0) \
+		return false
+
+bool is_valid_custom_header(const char *custom_header)
+{
+	const char *c;
+	int name_len;
+
+	if (custom_header == NULL)
+		return true;
+
+	// Disallow \r and \n
+	c = strchr(custom_header, '\r');
+	if (c != NULL)
+		return false;
+	c = strchr(custom_header, '\n');
+	if (c != NULL)
+		return false;
+
+	// Require a header name followed by :
+	c = strchr(custom_header, ':');
+	if (c == NULL)
+		return false;
+	name_len = c - custom_header;
+	if (name_len < 1)
+		return false;
+
+	// Disallow headers that we set
+	forbid_custom_header("User-Agent");
+	forbid_custom_header("Host");
+	forbid_custom_header("Accept");
+	forbid_custom_header("Content-Type");
+	forbid_custom_header("Transfer-Encoding");
+	forbid_custom_header("Content-Length");
```

Comment:
This construct feels foreign to C. I would rather see a list in a static `git_strarray` which we loop over. If we were to stick to macros, we should pass in the string we're validating instead of relying on a particular one being defined.

The macro should then be all-caps and have the `do { } while (0)` surrounding it, as it consists of more than one line.